### ISSUE TITLE
CF: Forcing TTL of 1 sec, when 300 was configured?

### ISSUE THREAD

**Issue by patschi** [StackExchange/dnscontrol#490]
While working on a tiny change in the cloudflare provider I've came across following code:

https://github.com/StackExchange/dnscontrol/blob/d782bb002bccac04d66321480725f3eb46044857/providers/cloudflare/cloudflareProvider.go#L240-L242

Does anyone know the exact use-case for this? Why should be the TTL forced to `1`, when it was set to `300 seconds` (5mins)? Does this has any particular reason?

*Edit*: I'm thinking about removing `|| rec.TTL == 300` as it doesn't make any sense to me and might cause unexpected behavior.

**Comment by tlimoncelli** [comment]
Yeah, we should add a comment to explain.

The DNS protocol doesn't permit a zero TTL. Therefore we use TTL=0 internally to mean "use the default for that zone". Most DNS providers use 0 to mean the same thing.

CloudFlare uses 0 to mean "set this to the automatic value", which is 300. When you retrieve the zone, you'll get 300, not 0.  This leads to a battle where every "dnscontrol push" will try to change the value to zero, but think that another update is needed because when it reads the records, it sees TTL=300.

I can't remember the justification for using TTL=1 but I do remember it was important. Maybe it was just a small value that wasn't zero?

Feel free to remove the `|| rec.TTL == 300` and run the integration tests.

Tom

**Comment by patschi** [comment]
> CloudFlare uses 0 to mean "set this to the automatic value", which is 300. When you retrieve the zone, you'll get 300, not 0.

Based on my research, that's not correct: Cloudflare uses a TTL of `1` to say "managed by cloudflare", according to their documentation: https://api.cloudflare.com/#dns-records-for-a-zone-properties (Quite well hidden at `DNS Records for a Zone -> DNS Records for a Zone properties -> Object definitions: Show definitions -> ttl`)

**Comment by tlimoncelli** [comment]
Based on my research, my memory is terrible :-)

Maybe we're both right.  The 0 `->` 1 translation is because dnscontrol uses 0 to mean "default" but we change it to 1 because we want the TTL to be managed by CF in that situation. (as you said)  The 300 `->` 1 translation might be to cover up very old domains or a change of the API?

Either way... feel free to play around with it and send a PR if removing that code works.  Either way, I'd love a PR that adds comments explaining what you found.

**Comment by captncraig** [comment]
I believe this is an artifact of several historical factors. The quality of them may be debatable again, but this is the best I can recall.

- 300 is a magic number that we chose at some point as the global ["default ttl"](https://github.com/StackExchange/dnscontrol/blob/921fa9849001d42494b23feadbe069d1dd46d916/models/dns.go#L9) if none is specified. It seemed a reasonable value at the time.
- Cloudflare also has a similar mode for "I don't care about ttls, take care of it for me", and you activate it by setting the ttl to `1`. They happened to use 300 for the value at the time (and now as far as I can tell).
- At some point we decided that if the user doesn't specify a ttl, the cloudflare managed mode is preferable to dnscontrol arbitrarily choosing 300 for those records. So we decided "if the ttl is 300, the likely explanation is that the user chose no ttl, and the default is in effect. So Let's use cloudflare's default instead of our own."
- We sacrifice for this the ability to actually specify 300 as a ttl for cloudflare records, but such is life.

The net result of merging your fix is that now I have a whole pile of corrections on all of our cloudflare domains that are all just ttl changes from 1 to 300. I still feel like using cloudflare automatic ttls is the preferable outcome, but don't have a super strong opinion either way. 

I'm inclined to back out that particular change for now, as I do consider it a breaking change, albeit minor.
